// ts-loader/examples/webpack2-gulp-react-flux-babel-karma/src/components/App.js
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var react_1 = require('react');
var GreetingStore_1 = require('../stores/GreetingStore');
var WhoToGreet_1 = require('./WhoToGreet');
var Greeting_1 = require('./Greeting');
var App = (function (_super) {
    __extends(App, _super);
    function App(props) {
        var _this = this;
        _super.call(this, props);
        this.onChange = function () {
            _this.setState(_this.getStateFromStores());
        };
        this.state = this.getStateFromStores();
    }
    App.prototype.componentWillMount = function () {
        this.eventSubscription = GreetingStore_1.default.addChangeListener(this.onChange);
    };
    App.prototype.componentWillUnmount = function () {
        this.eventSubscription.remove();
    };
    App.prototype.render = function () {
        var _a = this.state, greetings = _a.greetings, newGreeting = _a.newGreeting;
        return (<div className="container-fluid">
        <h1>Hello People!</h1>

        <WhoToGreet_1.default newGreeting={newGreeting}/>

        {greetings.map(function (g, index) { return <Greeting_1.default key={index} targetOfGreeting={g}/>; })}
      </div>);
    };
    App.prototype.getStateFromStores = function () {
        return GreetingStore_1.default.getState();
    };
    return App;
})(react_1.default.Component);
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = App;
